test/app-luatest/trigger_module_test.lua

```diff
@@ -0,0 +1,498 @@
+    local trigger = require('trigger')
+
+    local event = "my_event"
+    local name = "my_trigger"
+    local state = 0
+    local function handler(self, inc)
+        if inc == nil then inc = 1 end
+        state = state + inc
+    end
+
+    local function check_object(obj)
+        local ret_handler = trigger.set(event, name, obj)
+        t.assert_equals(ret_handler, obj)
+        state = 0
+        local new_state = math.random(1, 100)
+        trigger.call(event, new_state)
+        t.assert_equals(state, new_state)
+        trigger.del(event, name)
+    end
+
+    local a, b, c = generate_callable_objects(handler)
+
+    -- Callable table
+    check_object(a)
+    local mt = getmetatable(a)
+    mt.__metatable = "Private"
+    check_object(a)
+
+    -- Callable userdata
+    check_object(b)
+    mt = getmetatable(b)
+    mt.__metatable = "Private"
+    check_object(a)
+
+    -- Callable cdata
+    check_object(c)
+    -- Cdata metatable is already protected so a special check is not required
+end
+
+g.test_trigger_clear_from_trigger = function()
+    local trigger = require('trigger')
+    local event = "test_trigger_clear_event"
+    local order = {}
+    local f2 = function()
+        table.insert(order, 2)
+    end
+    local f1 = function()
+        table.insert(order, 1)
+        trigger.del(event, 'f2')
+        trigger.del(event, 'f1')
+    end
+    trigger.set(event, 'f2', f2)
+    trigger.set(event, 'f1', f1)
+    trigger.call(event)
+    t.assert_equals(order, {1}, "Correct trigger invocation when 1st trigger \
+                                 clears 2nd")
+end
+
+g.test_pairs = function()
+    local trigger = require('trigger')
+
+    local state = 0
+    local function handler(inc)
+        state = state + inc
+    end
+    local function handler_method(self, inc)
+        state = state + inc
+    end
+    local a, b, c = generate_callable_objects(handler_method)
+    local handlers = {handler, c, b, a, handler, c, a, handler, b, handler}
+    local event = 'event_name'
+    local trigger_prefix = 'trigger_name.'
+
+    for i = #handlers, 1, -1 do
+        trigger.set(event, trigger_prefix..tostring(i),  handlers[i])
+    end
+
+    local idx = 0
+    local sum = 0
+    for name, h in trigger.pairs(event) do
+        idx = idx + 1
+        sum = sum + idx
+        t.assert_equals(name, trigger_prefix .. tostring(idx))
+        t.assert_equals(h, handlers[idx])
+        h(idx)
+    end
+    t.assert_equals(idx, #handlers)
+    t.assert_equals(state, sum)
+    for i = 1, #handlers do
+        trigger.del(event, trigger_prefix .. tostring(i))
+    end
+end
+
+-- Check that a trigger deleted during pairs does not break anything and
+-- isn't yielded by pairs after its deletion.
+g.test_del_during_pairs = function()
+    local trigger = require('trigger')
+
+    local event = 'event'
+    local trigger_num = 100
+    local break_point = math.random(5, 95)
+
+    t.assert_equals(trigger.pairs(event), nil)
```

Comment:
The test fails here if called repeatedly:

```
$ luatest -v test/app-luatest/trigger_module_test.lua -r 2
Tarantool version is 3.0.0-alpha1-136-ga1596bdc3b74
Started on Wed Aug 23 11:48:30 2023
    app-luatest.trigger_module.test_simple ... (0.000s) Ok
    app-luatest.trigger_module.test_order_and_shared_argument ... (0.000s) Ok
    app-luatest.trigger_module.test_shared_handler ... (0.000s) Ok
    app-luatest.trigger_module.test_exceptions ... (0.000s) Ok
    app-luatest.trigger_module.test_any_callable_trigger ... (0.000s) Ok
    app-luatest.trigger_module.test_trigger_clear_from_trigger ... (0.000s) Ok
    app-luatest.trigger_module.test_pairs ... (0.000s) Ok
    app-luatest.trigger_module.test_del_during_pairs ... (0.001s) fail
...ntool/tarantool/test/app-luatest/trigger_module_test.lua:267: expected: nil, actual: function: 0x2d368310
    app-luatest.trigger_module.test_clear_event_during_pairs ... (0.001s) Ok
    app-luatest.trigger_module.test_invalid_args ... (0.001s) Ok
    app-luatest.trigger_module.test_nil_args ... (0.000s) Ok
    app-luatest.trigger_module.test_info ... (0.001s) Ok
    app-luatest.trigger_module.test_info_no_empty_events ... (0.000s) Ok
=========================================================

Failed tests:
-------------

1) app-luatest.trigger_module.test_del_during_pairs
...ntool/tarantool/test/app-luatest/trigger_module_test.lua:267: expected: nil, actual: function: 0x2d368310
stack traceback:
        ...ntool/tarantool/test/app-luatest/trigger_module_test.lua:267: in function 'app-luatest.trigger_module.test_del_during_pairs'
        ...
        [C]: in function 'xpcall'

Ran 13 tests in 0.006 seconds, 12 succeeded, 1 failed

=========================================================
Failed tests:

app-luatest.trigger_module.test_del_during_pairs
```

It also fails with `--shuffle all` occasionally. Please fix.

Reply:
I think that the iterator didn't get garbage collected, so it referenced the event and it was not deleted. That is why a new iterator was created from a non-NULL event. Now I use a unique name for the test to avoid such a situation. And this line was removed - now I check that pairs still work correctly but do not make iterations.